### PR TITLE
feat: Add feature to remove empty and tiny samples in `mteb.data_cleaning`

### PR DESCRIPTION
**Pull request by ayush1298**
Follow-up to #5107 (next step of #3672).
This PR add a feature to remove empty/short samples of different modalities from the task.

```
from mteb.data_cleaning import remove_short_texts, remove_small_images, remove_short_audio, remove_short_videos

remove_short_texts(task, min_length=1)                                   # empty / whitespace-only
remove_short_texts(task, min_length=2, length=lambda t: len(t.split()))  # count words
remove_small_images(task, min_size=32)                                   # width or height < 32px
remove_short_audio(task, min_seconds=0.5)
remove_short_videos(task, min_seconds=1.0)
```

- Removes a sample if any of its values is too small (e.g. either side of a pair). No default threshold, as it is task-dependent.
- Sizes are measured like the descriptive stats (`min_text_length`, `min_image_width/height`, `min_duration_seconds`), so the stats show what a threshold removes. Texts are stripped first; audio/video durations use helpers now shared with the stats.
- Retrieval: text is measured as the model reads it (title + text, query + instruction). A removed document takes its judgements with it, and queries without positives are dropped. Entries combining modalities (e.g. a page image with its text) are skipped.
- Same logic as `remove_duplicates`: returns a copy, name extends to `X (remove_duplicates, remove_short_texts)`.

**Tested** on NanoSCIDOCS, CodeRAG, IndicNLPNews, HUMECore17, DanRAG (t2it), WIT (images) and BeijingOpera (audio). After filtering, every stats minimum is at or above the threshold, and cleaned tasks run through `mteb.evaluate`.

Task | Filter | Result
-- | -- | --
WITT2IRetrieval | min_size=64 | 28 small-image documents removed; 29 queries dropped because they lost their only relevant document; text queries not measured. Smallest image height afterwards is 65px.
BeijingOpera | min_seconds=0.5 | 81 of 236 clips removed; shortest clip afterwards is 0.504s.

### CODE REVIEW

**Comment by ayush1298** [comment]
@Samoed can you have a look at this?

File: mteb/data_cleaning/_filters.py [hunk 1]
```diff
@@ -53,12 +60,13 @@ def _strip_whitespace(text: str) -> str:
     return text.strip()
 
 
-KeepIndicesFn = Callable[[Iterable[tuple[str, ...]]], list[int]]
-"""Given the comparable content of each row, return the (ascending) indices of the rows to keep.
+KeepIndicesFn = Callable[[Iterable[tuple[Any, ...]]], list[int]]
```
Comment (Samoed) [inline mteb/data_cleaning/_filters.py]: Sould this be still named `KeepIndicesFn`?

File: mteb/data_cleaning/_filters.py [hunk 2]
```diff
@@ -584,3 +652,227 @@ def remove_duplicates(
...
+    num_proc: int | None = None,
+) -> T:
+    """Remove samples with an image whose width or height is under `min_size` pixels, e.g. 1x1 placeholders.
+
+    Works like [`remove_short_texts`][mteb.data_cleaning.remove_short_texts], measuring images instead, the same way
+    as the `min_image_width` and `min_image_height` of the task's descriptive statistics.
+
+    Args:
+        task: The task to filter. It is not modified.
+        min_size: The smallest width and height, in pixels, an image may have.
```
Comment (Samoed) [inline mteb/data_cleaning/_filters.py]: maybe add min width/min height or make possible to pass function like in `remove_short_texts`

File: mteb/data_cleaning/_filters.py (hunk 1 above)
Comment (ayush1298) [reply mteb/data_cleaning/_filters.py]: I guess yes, because it still returns ascending indices of the rows to keep, though its input is changed. Maybe should we rename to `KeepRowsFn` or `SelectRowsFn`?

File: mteb/data_cleaning/_filters.py (hunk 2 above)
Comment (ayush1298) [reply mteb/data_cleaning/_filters.py]: Added size callable to `remove_small_images` to mirror `length` in `remove_short_texts` and now it can be used like: 

```
remove_small_images(task, min_size=32)  # shorter side, so at least 32px in both directions
remove_small_images(task, min_size=1024, size=lambda image: image.width * image.height)  # area
```

File: mteb/data_cleaning/_filters.py (hunk 1 above)
Comment (Samoed) [reply mteb/data_cleaning/_filters.py]: I think you can rename to `KeepRowsFn`

File: mteb/_create_dataloaders.py [hunk 3]
```diff
@@ -151,6 +151,31 @@ def _convert_conv_history_to_query(
...
+    """
+    if len(dataset) == 0:
+        return []
+    if prompt_type == PromptType.document:
+        dataset = dataset.map(_corpus_to_dict)
+    else:
+        if "instruction" in dataset.column_names:
+            dataset = _combine_queries_with_instruction_text(dataset)
+        if isinstance(dataset["text"][0], dict | list):
+            dataset = dataset.map(_convert_conv_history_to_query)
```
[the remark this answers is not in the document]
Comment (ayush1298) [reply mteb/_create_dataloaders.py]: done

File: mteb/data_cleaning/_filters.py (hunk 1 above)
Comment (ayush1298) [reply mteb/data_cleaning/_filters.py]: done

File: mteb/data_cleaning/_filters.py [hunk 4]
```diff
@@ -584,3 +652,232 @@ def remove_duplicates(
...
+    task: T,
+    *,
+    min_length: int,
+    length: Callable[[str], int] = len,
+    columns: Sequence[str] | None = None,
+    splits: Sequence[str] | None = None,
+    subsets: Sequence[HFSubset] | None = None,
+    num_proc: int | None = None,
+) -> T:
+    """Remove samples with a text shorter than `min_length`, which includes empty and whitespace-only texts.
```
Comment (KennethEnevoldsen) [inline mteb/data_cleaning/_filters.py:695]: too long - generally shorten

File: mteb/data_cleaning/_filters.py [hunk 5]
```diff
@@ -584,3 +652,232 @@ def remove_duplicates(
...
+    together, and a query includes its instruction. In characters, the default, they are counted like the
+    `min_text_length` of the task's descriptive statistics, apart from the stripped whitespace. A missing text
+    counts as empty.
+
+    Only text is measured. A retrieval document or query that combines text with an image, audio or video is kept,
+    as that content may carry what its text lacks. For a retrieval task, the relevance judgements of a removed
+    document go with it, and a query left without a relevant document is dropped. For a clustering task, short texts
+    are removed from within each set of texts, together with their labels.
+
+    The task passed in is left untouched; the cleaned copy is named after the filters applied to it, e.g.
```
Comment (KennethEnevoldsen) [inline mteb/data_cleaning/_filters.py:708]:
```suggestion
```

File: mteb/data_cleaning/_filters.py [hunk 6]
```diff
@@ -584,3 +652,232 @@ def remove_duplicates(
...
+    `MassiveIntentClassification (remove_short_texts)`.
```
Comment (KennethEnevoldsen) [inline mteb/data_cleaning/_filters.py:709]: should go in return statement

File: mteb/data_cleaning/_filters.py [hunk 7]
```diff
@@ -584,3 +652,232 @@ def remove_duplicates(
...
+    length: Callable[[str], int] = len,
+    columns: Sequence[str] | None = None,
+    splits: Sequence[str] | None = None,
+    subsets: Sequence[HFSubset] | None = None,
+    num_proc: int | None = None,
+) -> T:
+    """Remove samples with a text shorter than `min_length`, which includes empty and whitespace-only texts.
+
+    A sample is removed when any of its texts is too short, e.g. either sentence of a pair. Texts are measured as
+    the model reads them, after stripping surrounding whitespace: a retrieval document is its title and text
```
Comment (KennethEnevoldsen) [inline mteb/data_cleaning/_filters.py:698]: model? What model?

File: mteb/data_cleaning/_filters.py [hunk 8]
```diff
@@ -584,3 +652,232 @@ def remove_duplicates(
...
+        columns: The text columns to measure. Defaults to every text column of the task.
+        splits: The splits to filter. Defaults to every split of the dataset.
+        subsets: The Huggingface subsets to filter. Defaults to every loaded subset.
+        num_proc: Number of processes to use for loading and filtering the dataset.
+
+    Returns:
+        A copy of the task holding the filtered data.
+
+    Raises:
+        NotImplementedError: If `task` aggregates other tasks, or has no text.
```
Comment (KennethEnevoldsen) [inline mteb/data_cleaning/_filters.py:724]: Why would having no text raise a NotImplementedError? There is nothing to implement

File: mteb/data_cleaning/_filters.py [hunk 9]
```diff
@@ -584,3 +652,232 @@ def remove_duplicates(
...
+        subsets=subsets,
+        num_proc=num_proc,
+    )
+
+
+def remove_short_texts(
+    task: T,
+    *,
+    min_length: int,
+    length: Callable[[str], int] = len,
```
Comment (KennethEnevoldsen) [inline mteb/data_cleaning/_filters.py:689]:
```suggestion
    length_fn: Callable[[str], int] = len,
```

File: mteb/data_cleaning/_filters.py [hunk 10]
```diff
@@ -584,3 +652,232 @@ def remove_duplicates(
...
+def _shorter_side(image: Image.Image) -> int:
+    """The shorter of an image's width and height, so that a thin image counts as small. The default."""
+    return min(image.size)
+
+
+def remove_small_images(
+    task: T,
+    *,
+    min_size: int,
+    size: Callable[[Image.Image], float] = _shorter_side,
```
Comment (KennethEnevoldsen) [inline mteb/data_cleaning/_filters.py:757]:
```suggestion
    size_fn: Callable[[Image.Image], float] = _shorter_side,
```
general

File: mteb/data_cleaning/_filters.py [hunk 11]
```diff
@@ -584,3 +652,232 @@ def remove_duplicates(
...
+    min_seconds: float,
+    columns: Sequence[str] | None = None,
+    splits: Sequence[str] | None = None,
+    subsets: Sequence[HFSubset] | None = None,
+    num_proc: int | None = None,
+) -> T:
+    """Remove samples with an audio clip shorter than `min_seconds`.
+
+    Works like [`remove_short_texts`][mteb.data_cleaning.remove_short_texts], measuring audio instead, the same way
+    as the `min_duration_seconds` of the task's descriptive statistics.
```
Comment (KennethEnevoldsen) [inline mteb/data_cleaning/_filters.py:812]:
```suggestion
```
I am not sure this is needed (same above)

File: mteb/data_cleaning/_filters.py [hunk 12]
```diff
@@ -53,12 +60,13 @@ def _strip_whitespace(text: str) -> str:
     return text.strip()
 
 
-KeepIndicesFn = Callable[[Iterable[tuple[str, ...]]], list[int]]
```
Comment (KennethEnevoldsen) [inline mteb/data_cleaning/_filters.py:56]: why the name change?

File: mteb/data_cleaning/_filters.py [hunk 13]
```diff
@@ -247,13 +288,13 @@ def _count_values(dataset: Dataset, column: str, grouped: bool) -> int:
 def _apply_row_filter(
     dataset: Dataset,
     col_modalities: Mapping[str, Modalities],
-    keep_fn: KeepIndicesFn,
+    filter_: _Filter,
```
Comment (KennethEnevoldsen) [inline mteb/data_cleaning/_filters.py:291]: why the underscore a the end `filter_`

File: tests/test_data_cleaning.py [hunk 14]
```diff
@@ -659,3 +668,347 @@ def test_retrieval_compares_each_side_on_its_own_modality() -> None:
     assert cleaned_data["corpus"]["id"] == ["d1"]
     assert cleaned_data["queries"]["id"] == ["q1"]
     assert cleaned_data["relevant_docs"] == {"q1": {"d1": 1}}
+
+
+def test_retrieval_compares_conversations_by_their_turns() -> None:
```
Comment (KennethEnevoldsen) [inline tests/test_data_cleaning.py:673]: I am not sure what this tests - why does this ahve anything to do with turns?

File: tests/test_data_cleaning.py [hunk 15]
```diff
@@ -659,3 +668,347 @@ def test_retrieval_compares_each_side_on_its_own_modality() -> None:
...
+            }
+        ),
+        "relevant_docs": {"q1": {"d1": 1}, "q2": {"d2": 1}, "q3": {"d2": 1}},
+        "top_ranked": None,
+    }
+
+    task = remove_duplicates(task)
+
+    data = task.dataset[subset][split]
+    # only the repeated conversation is a duplicate, the others differ in their turns
```
Comment (KennethEnevoldsen) [inline tests/test_data_cleaning.py:698]: ?

File: tests/test_data_cleaning.py [hunk 16]
```diff
@@ -659,3 +668,347 @@ def test_retrieval_compares_each_side_on_its_own_modality() -> None:
...
+        "relevant_docs": {"q1": {"d1": 1}, "q2": {"d2": 1}},
+        "top_ranked": None,
+    }
+
+    task = remove_duplicates(task)
+
+    assert task.dataset[subset][split]["queries"]["id"] == ["q1", "q2"]
+
+
+def test_a_retrieval_task_in_the_older_layout_can_be_filtered() -> None:
```
Comment (KennethEnevoldsen) [inline tests/test_data_cleaning.py:727]: Why are we adding tests for duplicates in this PR?

File: tests/test_data_cleaning.py [hunk 17]
```diff
@@ -659,3 +668,347 @@ def test_retrieval_compares_each_side_on_its_own_modality() -> None:
...
+    task.data_loaded = True
+
+    cleaned = remove_duplicates(task)
+
+    data = cleaned.dataset["default"]["test"]
+    assert data["corpus"]["id"] == ["d1"]
+    assert data["relevant_docs"] == {"q1": {"d1": 1}}
+
+
+def _texts_task(texts: list[str | None]) -> MockClassificationTask:
```
Comment (KennethEnevoldsen) [inline tests/test_data_cleaning.py:742]:
```suggestion
def _create_texts_task(texts: list[str | None]) -> MockClassificationTask:
```

File: tests/test_data_cleaning.py [hunk 18]
```diff
@@ -659,3 +668,347 @@ def test_retrieval_compares_each_side_on_its_own_modality() -> None:
...
+
+
+@pytest.mark.parametrize(
+    ("min_length", "expected"),
+    [
+        (1, ["hi", " hello there "]),
+        (3, [" hello there "]),
+    ],
+)
+def test_remove_short_texts_removes_texts_under_min_length(
```
Comment (KennethEnevoldsen) [inline tests/test_data_cleaning.py:758]:
```suggestion
@pytest.mark.parametrize(
    ("min_length", "input", "expected"),
    [
        (1, ["", "   ", None, "hi", " hello there "], ["hi", " hello there "]),
        (3, ["", "   ", None, "hi", " hello there "], [" hello there "]),
    ],
)
def test_remove_short_texts_removes_texts_under_min_length(
```
that allows for easier extension and it is easier to know why the expected output makes sense

File: mteb/data_cleaning/_filters.py (hunk 7 above)
Comment (ayush1298) [reply mteb/data_cleaning/_filters.py]: rewrite whole thing in a better and short words

File: mteb/data_cleaning/_filters.py (hunk 13 above)
Comment (ayush1298) [reply mteb/data_cleaning/_filters.py]: `filter` is a Python builtin, and `PEP8` says to append a single trailing underscore rather than shadow it

File: mteb/data_cleaning/_filters.py (hunk 12 above)
Comment (ayush1298) [reply mteb/data_cleaning/_filters.py]: `KeepIndicesFn` described only the return value (indices). Now that input rows can arrive as raw images or audio, not just strings, so the name change

File: mteb/data_cleaning/_filters.py (hunk 8 above)
Comment (ayush1298) [reply mteb/data_cleaning/_filters.py]: change it to `ValueError`, so its now something like: 

```
NotImplementedError: If `task` aggregates other tasks, which hold the data instead.
ValueError: If `task` has no text, or if `splits` and `subsets` together match none of its splits.
KeyError: If `columns` names a column that is not a text column of the task.
```

File: mteb/data_cleaning/_filters.py (hunk 11 above)
Comment (ayush1298) [reply mteb/data_cleaning/_filters.py]: removed this at all places

File: tests/test_data_cleaning.py (hunk 16 above)
Comment (ayush1298) [reply tests/test_data_cleaning.py]: they're for some of the bugs that this PR fixes like in `_retrieval_texts` which `duplicates` also used

File: tests/test_data_cleaning.py (hunk 14 above)
Comment (ayush1298) [reply tests/test_data_cleaning.py]: added a comment to make it clear, the name of test causes confusion

File: tests/test_data_cleaning.py (hunk 15 above)
Comment (ayush1298) [reply tests/test_data_cleaning.py]: clarified comment
